application/src/main/java/org/thingsboard/server/service/alarm/rule/DefaultTbAlarmRuleStateService.java

```diff
@@ -0,0 +1,517 @@
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.UUID;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.locks.Lock;
+import java.util.stream.Collectors;
+
+import static org.thingsboard.server.common.data.DataConstants.INTERNAL_QUEUE_NAME;
+import static org.thingsboard.server.common.data.plugin.ComponentLifecycleEvent.DELETED;
+
+@Slf4j
+@Service
+@TbRuleEngineComponent
+@RequiredArgsConstructor
+public class DefaultTbAlarmRuleStateService extends AbstractPartitionBasedService<EntityId> implements TbAlarmRuleStateService {
+
+    private final TbAlarmRuleContext ctx;
+    private final AlarmRuleService alarmRuleService;
+    private final AlarmRuleEntityStateStore stateStore;
+    //    private final RelationService relationService;
+    private final TbDeviceProfileCache deviceProfileCache;
+    private final TbAssetProfileCache assetProfileCache;
+
+    private final Map<EntityId, EntityState> entityStates = new ConcurrentHashMap<>();
+    private final Map<AlarmRuleId, Set<EntityId>> myEntityStateIdsPerAlarmRule = new ConcurrentHashMap<>();
+    private final Map<TenantId, Set<EntityId>> myEntityStateIds = new ConcurrentHashMap<>();
+    private Cache<TenantId, ConcurrentHashMap<AlarmRuleId, AlarmRule>> rulesCache;
+
+    @PostConstruct
+    protected void init() {
+        super.init();
+        rulesCache = Caffeine.newBuilder()
+                .expireAfterAccess(24, TimeUnit.HOURS)
+                .maximumSize(1024).build();
+
+        scheduledExecutor.scheduleAtFixedRate(this::harvestAlarms, 1, 1, TimeUnit.MINUTES);
+    }
+
+    protected ServiceType getServiceType() {
+        return ServiceType.TB_RULE_ENGINE;
+    }
+
+    @Override
+    protected String getServiceName() {
+        return "Alarm Rule State";
+    }
+
+    @Override
+    protected String getSchedulerExecutorName() {
+        return "alarm-rule-state-scheduler";
+    }
+
+    @Override
+    protected Map<TopicPartitionInfo, List<ListenableFuture<?>>> onAddedPartitions(Set<TopicPartitionInfo> addedPartitions) {
+        AtomicInteger addedEntityStates = new AtomicInteger(0);
+        addedPartitions.stream().map(tpi -> tpi.getTenantId().orElse(TenantId.SYS_TENANT_ID)).distinct().forEach(tenantId -> {
+            List<PersistedEntityState> states = stateStore.getAll(tenantId);
+            addedEntityStates.addAndGet(states.size());
+            states.forEach(ares -> {
+                try {
+                    EntityId entityId = ares.getEntityId();
+                    if (!entityStates.containsKey(entityId)) {
+                        createEntityState(ares.getTenantId(), entityId, ares);
+                    }
+                } catch (Exception e) {
+                    log.warn("Failed to create entity state [{}]!", e.getMessage());
```

Comment:
I think it will be hard to investigate some failure having only the error message, let's print the stacktrace in case errors shouldn't normally occur here